cranelift/codegen/src/isa/x64/inst.isle

```diff
@@ -4282,40 +4282,115 @@
       )
        (require (or (= ty 32) (= ty 64)))
 )
-(rule (x64_add_mem $I8 addr val) (x64_addb_mr_mem addr val))
-(rule (x64_add_mem $I16 addr val) (x64_addw_mr_mem addr val))
-(rule (x64_add_mem $I32 addr val) (x64_addl_mr_mem addr val))
-(rule (x64_add_mem $I64 addr val) (x64_addq_mr_mem addr val))
-
-(decl x64_sub_mem (Type Amode Gpr) SideEffectNoResult)
-(rule (x64_sub_mem $I8 addr val) (x64_subb_mr_mem addr val))
-(rule (x64_sub_mem $I16 addr val) (x64_subw_mr_mem addr val))
-(rule (x64_sub_mem $I32 addr val) (x64_subl_mr_mem addr val))
-(rule (x64_sub_mem $I64 addr val) (x64_subq_mr_mem addr val))
-
-(decl x64_and_mem (Type Amode Gpr) SideEffectNoResult)
+
+;; `add mem, reg`
+(rule 0 (x64_add_mem $I8 addr val) (x64_addb_mr_mem addr val))
+(rule 0 (x64_add_mem $I16 addr val) (x64_addw_mr_mem addr val))
+(rule 0 (x64_add_mem $I32 addr val) (x64_addl_mr_mem addr val))
+(rule 0 (x64_add_mem $I64 addr val) (x64_addq_mr_mem addr val))
+
+;; `add mem, imm`
+(rule 1 (x64_add_mem $I8 addr (u8_from_iconst val)) (x64_addb_mi_mem addr val))
+(rule 1 (x64_add_mem $I16 addr (u16_from_iconst val)) (x64_addw_mi_mem addr val))
+(rule 1 (x64_add_mem $I32 addr (u32_from_iconst val)) (x64_addl_mi_mem addr val))
+(rule 1 (x64_add_mem $I64 addr (i64_from_iconst val))
+      (if-let val2 (i32_try_from_i64 val))
+      (x64_addq_mi_sxl_mem addr val2))
+(rule 2 (x64_add_mem $I32 addr (i32_from_iconst (i8_try_from_i32 val)))
+      (x64_addl_mi_sxb_mem addr val))
+(rule 2 (x64_add_mem $I64 addr (i64_from_iconst val))
+      (if-let val2 (i8_try_from_i64 val))
+      (x64_addq_mi_sxb_mem addr val2))
```

Comment:
Would it make sense to add `i8_from_iconst` to make these two cases one-liners too?

Reply:
Similarly, this also needs to be fallible